### PR TITLE
[vcpkg] Add initial versioning documentation

### PR DESCRIPTION
**Pull request by ras0219**
This PR brings versioning into a near-ship state:

1. Adds reference documentation for the experimental user-facing features of versioning.
2. Renames `$x-default-baseline` to `builtin-baseline` and adds proper parsing support
3. Improves error messages throughout the versioning experience to help guide users to success
4. Improves performance of versioning to avoid all git manipulation in the happy path.
5. Uses `git archive` to achieve high performance git tree checkouts.
6. Improved test coverage of versioning
7. Users without any versioning fields will not experience any behavior changes with the `versions` feature flag enabled (+test).
8. Remove `version=`.

Two issues not addressed in this PR:
1. Emit warnings if the user uses overrides or constraints in a top-level manifest and doesn't have a baseline set (the user won't get versioning behavior in this case and that could be surprising)
2. Overrides currently uses schemed field names, however those schemes have no semantic meaning.

The initial spec is in #11758. Convenience link to the new docs is: https://github.com/ras0219/vcpkg/blob/dev/roschuma/versions/docs/users/versioning.md.

For local experiments, you will need to pull in versioning files:
```
git pull https://github.com/ras0219/vcpkg dev/roschuma/version-files
```
### Example:
As an example, you can try:
```json
{
    "name": "example",
    "version-date": "2020-01-01",
    "builtin-baseline": "334ba25e577b83331d5cfeb14d65768553ea1d9e",
    "dependencies": [
        {
            "$comment": "Note that this old rapidjson uses version-string",
            "name": "rapidjson",
            "version>=": "2019-06-28"
        },
        "zlib"
    ],
    "overrides": [
        {
            "name": "rapidjson",
            "version-string": "2019-06-28"
        }
    ]
}
```
and in that directory
```
vcpkg install --feature-flags=manifests,versions
vcpkg list --feature-flags=manifests,versions
```
Observing the result of `2019-06-28` and `1.2.11#9` for `rapidjson` and `zlib` respectively. If you didn't pull in the version files, you should see an error message like:
```
Failed at [/workspaces/vcpkg/toolsrc/src/vcpkg/registries.cpp(296)] with message:
Error: Couldn't find explicitly specified baseline `"334ba25e577b83331d5cfeb14d65768553ea1d9e"` in the baseline file, and there was no baseline at that commit or the commit didn't exist.
Error: while checking out baseline '334ba25e577b83331d5cfeb14d65768553ea1d9f:port_versions/baseline.json':
fatal: Path 'port_versions/baseline.json' does not exist in '334ba25e577b83331d5cfeb14d65768553ea1d9f'

This may be fixed by updating vcpkg to the latest master via `git pull`.
The current commit is "efe40c8c882c2bf3125b27e10294089f94e7b87b"
```
or
```
Failed at [/workspaces/vcpkg/toolsrc/src/vcpkg/install.cpp(877)] with message:
Error: no version entry for rapidjson at version 2019-06-28.
We are currently using the version in the ports tree (2020-09-14), since no rapidjson.json was found in /port_versions.
```

### CODE REVIEW

File: docs/users/versioning.md [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+Overrides are not considered from ports consumed as a dependency.
+
+## Example top-level manifest:
+```json
+{
+    "name": "example",
+    "version": "1.0",
+    "builtin-baseline": "a14a6bcb27287e3ec138dba1b948a0cdbc337a3a",
+    "dependencies": [
+        { "name": "zlib", "version>=": "1.2.11", "port-version": 8 },
```
Comment (BillyONeal) [inline docs/users/versioning.md:72]: Does the >= apply to the `port-version` here too?

Comment (ras0219) [reply docs/users/versioning.md:72]: `port-version` is always `>=` (and can't be specified without `version>=`). I could see a rename to `port-version>=` @vicroms?

Comment (BillyONeal) [reply docs/users/versioning.md:72]: At risk of repainting this bike shed again could we do:

```
{ "name": "zlib", "version>=": "1.2.11#8" }
```

?

**Comment by ras0219** [comment]
I've added support for `{ "name": "zlib", "version>=": "1.2.11#8" }` via the comment banners `SUPPORT_HASH_SYNTAX`; once the bikeshed has been painted, we can delete any offending code.

File: scripts/azure-pipelines/end-to-end-tests-dir/registries.ps1 [hunk 2]
```diff
@@ -90,37 +93,8 @@ $vcpkgJson = @{
     )
 }
 
-$manifestDir = "$TestingRoot/builtin-registry-test-manifest-dir"
```
Comment (BillyONeal) [inline scripts/azure-pipelines/end-to-end-tests-dir/registries.ps1:93]: It doesn't seem like you intended to delete this?

Comment (BillyONeal) [reply scripts/azure-pipelines/end-to-end-tests-dir/registries.ps1:93]: Oh, I see, it got moved into foreach loop below.

File: toolsrc/src/vcpkg/sourceparagraph.cpp [hunk 3]
```diff
@@ -1271,15 +1271,18 @@ namespace vcpkg
...
+                    if (dep.constraint.port_version != 0)
+                    {
+                        dep_obj.insert(DependencyDeserializer::VERSION_GE,
+                                       Json::Value::string(
+                                           Strings::concat(dep.constraint.value, '#', dep.constraint.port_version)));
+                    }
+                    else
+                    {
+                        dep_obj.insert(DependencyDeserializer::VERSION_GE, Json::Value::string(dep.constraint.value));
+                    }
```
Comment (BillyONeal) [inline toolsrc/src/vcpkg/sourceparagraph.cpp:1285]: Suggest DRYing this out:

```suggestion
                    string versionConstraintStr;
                    if (dep.constraint.port_version == 0) {
                        versionConstraintStr = dep.constraint.value;
                     } else {
                        versionConstraintStr = Strings::concat(dep.constraint.value, '#', dep.constraint.port_version);
                    }

                    dep_obj.insert(DependencyDeserializer::VERSION_GE, Json::Value::string(std::move(versionConstraintStr)));
```

Comment (BillyONeal) [reply toolsrc/src/vcpkg/sourceparagraph.cpp:1285]: Or even something like:

```
                    string versionConstraintStr = dep.constraint.value;
                    if (dep.constraint.port_version != 0) {
                        versionConstraintStr.push_back('#');
                        versionConstraintStr.append(std::to_string(dep.constraint.port_version));
                    }
```

**Review by BillyONeal — APPROVED** [review]

File: scripts/azure-pipelines/end-to-end-tests-dir/registries.ps1 (hunk 2 above)
Comment (ras0219) [reply scripts/azure-pipelines/end-to-end-tests-dir/registries.ps1:93]: This test worked out to be a complete duplicate of the existing test in `versions.ps1`, so I added a loop around that test which enables the registries flag.

**Pull request merged**